src/scores/probability/fss_impl.py

```diff
@@ -0,0 +1,528 @@
+    """
+    Calculates the Fractions Skill Score (FSS) for a given forecast and observed
+    2-D field.
+
+    The FSS is computed by counting the squared sum of forecast and observation
+    events in a given window. This is repeated over all possible window
+    positions that can fit in the input arrays. A common method to do this is
+    via a sliding window.
+
+    While the various compute backends may have their own implementations; most
+    of them use some form of prefix sum algorithm to compute this quickly.
+
+    For 2-D fields this data structure is known as the "Summed Area
+    Table"<sup>1.</sup>.
+
+    Once the squared sums are computed, the final FSS value can be derived by
+    accumulating the squared sums<sup>2.</sup>.
+
+
+    The caller is responsible for making sure the input fields are in the 2-D
+    spatial domain. (Although it should work for any `np.array` as long as it's
+    2D, and `window` is appropriately sized.)
+
+    Args:
+        fcst: An array of forecasts
+        obs: An array of observations (same spatial shape as `fcst`)
+        threshold: A scalar to compare `fcst` and `obs` fields to generate a
+            binary "event" field.
+        window: A pair of positive integers `(height, width)` of the sliding
+            window; the window dimensions must be greater than 0 and fit within
+            the shape of `obs` and `fcst`.
+        compute_method: currently only supports `FssComputeMethod.NUMPY`
+            see: :py:class:`FssComputeMethod`
+
+    Returns:
+        A float representing the accumulated FSS.
+
+    Raises:
+        DimensionError: Various errors are thrown if the input dimensions
+            do not conform. e.g. if the window size is larger than the input
+            arrays, or if the the spatial dimensions of the input arrays do
+            not match.
+
+    References:
+        1. https://en.wikipedia.org/wiki/Summed-area_table
+        2. https://www.researchgate.net/publication/269222763_Fast_calculation_of_the_Fractions_Skill_Score
+    """
+    fss_backend = FssBackend.get_compute_backend(compute_method)
+    fb_obj = fss_backend(fcst, obs, threshold=threshold, window=window, zero_padding=zero_padding)
+    fss_score = fb_obj.compute_fss()
+
+    return fss_score
+
+
+def _aggregate_fss_decomposed(fss_d: FssDecomposed) -> np.float64:
+    """
+    Aggregates the results of decomposed fss scores.
+    """
+    # can't do ufuncs over custom void types currently...
+    fcst_sum = 0.0
+    obs_sum = 0.0
+    diff_sum = 0.0
+
+    if isinstance(fss_d, np.ndarray):
+        l = fss_d.size
+        if l < 1:
+            return 0.0
+        with np.nditer(fss_d) as it:
+            for elem in it:
+                (fcst_, obs_, diff_) = elem.item()
+                fcst_sum += fcst_ / l
+                obs_sum += obs_ / l
+                diff_sum += diff_ / l
+    else:
+        (fcst_sum, obs_sum, diff_sum) = fss_d
+
+    fss = 0.0
+    denom = obs_sum + fcst_sum
```

Comment:
What happens if you supply two empty grids and this sum is zero?

Reply:
Just so long as it doesn't div by zero and throw an error. Good to know it's in the tests. We should probably do a walkthrough of this implementation at some point, so make some notes somewhere about these choices. Perhaps also put such notes into the module documentation so that they are available to future readers.